File: TypeScript/ExampleTests/common/SimpleClassString.ts

import * as tokens from "./tokens.js";

export class SimpleClassString {
  static readonly DNC_STRING = "?DNC?";

  anInt: string;
  aString: string;

  constructor(anInt: string = "", aString: string = "") {
    this.anInt = anInt;
    this.aString = aString;
  }

  /** Builds from the text form, e.g. Money as "25 USD". */
  static fromText(text: string): SimpleClassString {
    const parts = tokens.require_(text, 2, "SimpleClass");
    return new SimpleClassString(
      parts[0],
      parts[1]
    );
  }

  static fromList(values: Iterable<string>): SimpleClassString {
    const v = Array.from(values);
    const r = new SimpleClassString();
    r.anInt = v[0] ?? "";
    r.aString = v[1] ?? "";
    return r;
  }

  toString(): string {
    return tokens.token(this.anInt) + " " + tokens.token(this.aString);
  }

  equals(other: SimpleClassString): boolean {
    return (this.anInt === SimpleClassString.DNC_STRING || other.anInt === SimpleClassString.DNC_STRING || this.anInt === other.anInt)
      && (this.aString === SimpleClassString.DNC_STRING || other.aString === SimpleClassString.DNC_STRING || this.aString === other.aString);
  }

  static equalLists(a: SimpleClassString[], b: SimpleClassString[]): boolean {
    if (a.length !== b.length) return false;
    return a.every((row, i) => row.equals(b[i]!));
  }
}

File: CSharp/SpecStudioExamplesTestsCSharp/common/SimpleClassString.cs
namespace SpecStudioExamplesTestsCSharp.common
{
using production;

    public class SimpleClassString
    {
        public string anInt;
        public string aString;

        public SimpleClassString(string anInt, string aString)
        {
            this.anInt = anInt;
            this.aString = aString;
        }

        /// <summary>Builds from the text form, e.g. Money as "25 USD".</summary>
        public static SimpleClassString FromText(string text)
        {
            var parts = Tokens.Require(text, 2, "SimpleClass");
            return new SimpleClassString(parts[0], parts[1]);
        }

        public SimpleClassString(string text)
        {
            var parsed = FromText(text);
            this.anInt = parsed.anInt;
            this.aString = parsed.aString;
        }

        public SimpleClassTyped ToSimpleClassTyped()
        {
            return new SimpleClassTyped(
                int.Parse(this.anInt),
                this.aString
            );
        }

        public override string ToString()
        {
            return Tokens.Token(anInt) + " " + Tokens.Token(aString);
        }

        const string DNCString = "?DNC?";

        public override bool Equals(object? obj)
        {
            if (obj is not SimpleClassString other) return false;
            return (DNCString == this.anInt || DNCString == other.anInt || object.Equals(this.anInt, other.anInt))
                && (DNCString == this.aString || DNCString == other.aString || object.Equals(this.aString, other.aString));
        }

        public override int GetHashCode()
        {
            var h = new System.HashCode();
            h.Add(this.anInt);
            h.Add(this.aString);
            return h.ToHashCode();
        }
    }
}
